node/nodeLoadAccounts.go

```diff
@@ -0,0 +1,152 @@
+package node
+
+import (
+	"encoding/hex"
+	"errors"
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/data/api"
+	"github.com/ElrondNetwork/elrond-go/state"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+type blockInfo struct {
+	nonce    uint64
+	hash     []byte
+	rootHash []byte
+}
+
+func (info *blockInfo) toApiResource() api.BlockInfo {
+	return api.BlockInfo{
+		Nonce:    info.nonce,
+		Hash:     hex.EncodeToString(info.hash),
+		RootHash: hex.EncodeToString(info.rootHash),
+	}
+}
+
+func (n *Node) loadUserAccountHandlerByAddress(address string, options api.AccountQueryOptions) (state.UserAccountHandler, api.BlockInfo, error) {
+	pubKey, err := n.decodeAddressToPubKey(address)
+	if err != nil {
+		return nil, api.BlockInfo{}, err
+	}
+
+	return n.loadUserAccountHandlerByPubKey(pubKey, options)
+}
+
+func (n *Node) loadUserAccountHandlerByPubKey(pubKey []byte, options api.AccountQueryOptions) (state.UserAccountHandler, api.BlockInfo, error) {
+	blockInfo, err := n.getBlockInfoGivenOptions(options)
+	if err != nil {
+		return nil, api.BlockInfo{}, err
+	}
+
+	account, err := n.loadUserAccountOnRootHash(pubKey, blockInfo.rootHash)
+	if err != nil {
+		return nil, api.BlockInfo{}, err
+	}
+
+	userAccount, ok := n.castAccountToUserAccount(account)
+	if !ok {
+		return nil, api.BlockInfo{}, ErrCannotCastAccountHandlerToUserAccountHandler
+	}
+
+	return userAccount, blockInfo.toApiResource(), nil
+}
+
+func (n *Node) loadUserAccountOnRootHash(pubKey []byte, rootHash []byte) (vmcommon.AccountHandler, error) {
+	repository, err := n.getAccountsRepository()
+	if err != nil {
+		return nil, err
+	}
+
+	account, err := repository.GetExistingAccount(pubKey, rootHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return account, nil
+}
+
+func (n *Node) loadAccountCode(codeHash []byte, options api.AccountQueryOptions) ([]byte, api.BlockInfo) {
+	blockInfo, err := n.getBlockInfoGivenOptions(options)
+	if err != nil {
+		return nil, api.BlockInfo{}
+	}
+
+	repository, err := n.getAccountsRepository()
+	if err != nil {
+		return nil, api.BlockInfo{}
+	}
+
+	code := repository.GetCode(codeHash, blockInfo.rootHash)
+	return code, blockInfo.toApiResource()
+}
+
+func (n *Node) getAccountsRepository() (state.AccountsRepository, error) {
+	// Question for review: do we still require the special lazy initialization logic (used on some components in node.go)?
```

Comment:
I would say no. we use other components inside node without checking if they are nil. However, the `options` approach doesn't guarantee us that all the members are filled

Reply:
Simplified.